src/Orleans.GrainDirectory.Redis/RedisGrainDirectory.cs

```diff
@@ -89,11 +89,24 @@ public async Task<GrainAddress> Register(GrainAddress address)
 
         public async Task Unregister(GrainAddress address)
         {
-            var value = JsonConvert.SerializeObject(address);
+            var value = JsonSerializer.Serialize(address);
 
             try
             {
-                var result = (int) await this.database.ScriptEvaluateAsync(this.deleteScript, new { key = GetKey(address.GrainId), val = JsonConvert.SerializeObject(address) });
+                if (address.MembershipVersion == MembershipVersion.MinValue)
+                {
+                    var fullAddress = await Lookup(address.GrainId);
+                    if (!fullAddress.Equals(address))
```

Comment:
It's unclear that `GrainAddress.Equals` does not compare the `MembershipVersion` property

Reply:
On `SiloAddress`, we have a `bool Matches(SiloAddress other)` method to serve the purpose of "same identity but not identical" - maybe we should add that on `GrainAddress`